src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -71,33 +81,67 @@ public class TemplateBinder
             _pool = pool;
             _pattern = pattern;
             _defaults = defaults;
+            _requiredKeys = requiredKeys?.ToArray() ?? Array.Empty<string>();
+
+            for (var i = 0; i < _requiredKeys.Length; i++)
+            {
+                var requiredKey = _requiredKeys[i];
+                if (_pattern.GetParameter(requiredKey) != null)
+                {
+                    throw new InvalidOperationException(
+                        $"The parameter {requiredKey} can not be used as a required key since it appears as " +
+                        $"a parameter in the route pattern.");
+                }
+            }
 
             // Any default that doesn't have a corresponding parameter is a 'filter' and if a value
             // is provided for that 'filter' it must match the value in defaults.
-            _filters = new RouteValueDictionary(_defaults);
+            var filters = new RouteValueDictionary(_defaults);
             foreach (var parameter in _pattern.Parameters)
             {
-                _filters.Remove(parameter.Name);
+                filters.Remove(parameter.Name);
             }
+
+            _filters = filters.ToArray();
+            
+            _slots = AssignSlots(_pattern, _filters);
         }
 
         // Step 1: Get the list of values we're going to try to use to match and generate this URI
         public TemplateValuesResult GetValues(RouteValueDictionary ambientValues, RouteValueDictionary values)
         {
-            return GetValues(ambientValues: ambientValues, explicitValues: values, requiredKeys: null);
-        }
-
-        internal TemplateValuesResult GetValues(
-            RouteValueDictionary ambientValues,
-            RouteValueDictionary explicitValues,
-            IEnumerable<string> requiredKeys)
-        {
-            var context = new TemplateBindingContext(_defaults);
-
-            var canCopyParameterAmbientValues = true;
+            // Make a new copy of the slots array, we'll use this as 'scratch' space.
+            var slots = new KeyValuePair<string, object>[_slots.Length];
```

Comment:
Could this be stackalloced?

Reply:
Add a comment that it will be used to back a RVD

Reply:
You can't stack this anyway since it contains reference types 😁